### PR TITLE
Revamp macro system to fix scope extrusion bugs

### LINKED ISSUE #535: simple macro with \section fails to compile... unless  written in xml flavour
when using the folowing macro:
```
\define[command=mois]{\section[numbering=no]{\process}}
```
sile fails by:
```
Error detected:
/usr/local/share/sile/core/inputs-common.lua:26: bad argument #1 to 'ipairs' (table expected, got nil)

```
[mois.txt](https://github.com/simoncozens/sile/files/2173176/mois.txt)

in xml the compilation works nice.
```
<define command="mois">
<section numbering="no"><process/></section>
</define>

<mois>Juillet</mois>
```

### PR DESCRIPTION
**Pull request by OlivierNicole**
The bug in #535 is due to a `\process` from a macro being evaluated after the macro finished executing.

A possible solution to that is to replace every occurrence of `\process` by the argument before processing the body. However, macros can also take functions as their arguments, so this works if `SILE.process` accepts ASTs containing functions.

So this fixes #535. The implementation also allows for several occurrences of `\process` in a macro.

### CODE REVIEW

**Comment by alerque** [comment]
Hey look at you going from fixing a blurb in the docs to open heart surgery!

**Comment by OlivierNicole** [comment]
> Did you do any benchmarknig for how this tree walking on the contents of ever command affects performance?

I did not. It’s true that every macro application processes the macro body twice: once for replacing `\process`, and once for processing, as opposed to only once (for processing) currently. The overall cost remains linear in the number of nodes of the expanded AST, but I agree that some experimental evaluation would be necessary. Are there *big* documents with a lot of macros for us to play with?

File: core/baseclass.lua [hunk 1]
```diff
@@ -40,30 +40,53 @@ SILE.doTexlike = function (doc)
...
+  end
+end
+
 -- Need the \define command *really* early on in SILE startup
-local macroStack = {}
 SILE.registerCommand("define", function (options, content)
   SU.required(options, "command", "defining command")
+  if type(content) == "function" then
+    -- A macro defined as a function can take no argument, so we register
+    -- it as-is.
```
[the remark this answers is not in the document]
Comment (OlivierNicole) [reply core/baseclass.lua:61]: Well, it theoretically can. But defining a macro as a function means doing `SILE.call("define", …)`, e.g.:
```lua
SILE.call("define", {command = "mymacro"}, function()
  -- ...
  SILE.call("process")
  -- ...
end)
```
which is a very strange way of doing:
```
\define[command=mymacro]{ ... \process ... }
```
In fact, I don't even see why a user would want to call define from Lua and give it a function, but it is supported in order not to break `packages/xmltricks` which does it.

**Comment by alerque** [comment]
> > Did you do any benchmarknig for how this tree walking on the contents of ever command affects performance?
>
> I did not. It’s true that every macro application processes the macro body twice: once for replacing `\process`, and once for processing, as opposed to only once (for processing) currently.

I was less worried about the cost of processing macro invocations as somehow walking the trees of everything.

I ran some bencharks (`make -j1 benchmarks HEADSHA=47a0af8 BASESHA=2c13d78`)and the difference is entirely in the noise. For 2c13d78 where your branched from:

```json
{ "real": "38.328", "user": "37.151", "sys": "0.918" }
```

For 47a0af8:

```json
{ "real": "38.744", "user": "37.641", "sys": "0.882" }
```

I think all I'd do at this point is change that error to a warning and let people be recursively stupid if they want to be.

File: core/baseclass.lua (hunk 1 above)
[the remark this answers is not in the document]
Comment (alerque) [reply core/baseclass.lua:61]: Point taken, `SILE.call("define")` doesn't make any sense when we have `SILE.registerCommand()` ... speaking of which I've been meaning to bring up normalizing those, but that's a job for later.

File: core/baseclass.lua [hunk 2]
```diff
@@ -40,30 +40,53 @@ SILE.doTexlike = function (doc)
...
+    SILE.registerCommand(options["command"], content)
+    return
+  elseif options.command == "process" then
+    SU.error("Defining a macro named `process`? Yeah, that won't go well.")
```
[the remark this answers is not in the document]
Comment (alerque) [reply core/baseclass.lua:65]:
```suggestion
    SU.warn("Did you mean to re-definine the `\process` macro? That probably won't go well.")
```

**Pull request merged**